src/main/scala/org/broadinstitute/dsde/firecloud/EntityService.scala

```diff
@@ -220,36 +220,48 @@ class EntityService(rawlsDAO: RawlsDAO, importServiceDAO: ImportServiceDAO, cwds
+      getWorkspaceId(workspaceNamespace, workspaceName, userInfo) flatMap { workspaceId =>
+        val bucketToWrite = GcsBucketName(FireCloudConfig.Cwds.bucket)
+        val fileToWrite = GcsObjectName(s"to-cwds/${workspaceId}/${java.util.UUID.randomUUID()}.json")
+        val gcsPath = writeDataToGcs(bucketToWrite, fileToWrite, dataBytes)
+        val importRequest = getImportRequest(gcsPath, isUpsert)
+        importToCWDS(workspaceNamespace, workspaceName, workspaceId, userInfo, importRequest)
+      }
     } else {
+      val bucketToWrite = GcsBucketName(FireCloudConfig.ImportService.bucket)
+      val fileToWrite = GcsObjectName(s"incoming/${java.util.UUID.randomUUID()}.json")
+      val gcsPath = writeDataToGcs(bucketToWrite, fileToWrite, dataBytes)
+      val importRequest = getImportRequest(gcsPath, isUpsert)
       importServiceDAO.importJob(workspaceNamespace, workspaceName, importRequest, isUpsert)(userInfo)
     }
   }
 
-  private def importToCWDS(workspaceNamespace: String, workspaceName: String, userInfo: UserInfo, importRequest: AsyncImportRequest
+  private def writeDataToGcs(bucket: GcsBucketName, objectName: GcsObjectName, data: Array[Byte]): String = {
+    val insertedObject = googleServicesDAO.writeObjectAsRawlsSA(bucket, objectName, data)
+    s"gs://${insertedObject.bucketName.value}/${insertedObject.objectName.value}"
+  }
+
+  private def getImportRequest(gcsPath: String, isUpsert: Boolean): AsyncImportRequest = {
+    AsyncImportRequest(gcsPath, FILETYPE_RAWLS, Some(ImportOptions(None, Some(isUpsert))))
+  }
+
+  private def getWorkspaceId(workspaceNamespace: String, workspaceName: String, userInfo: UserInfo): Future[String] = {
+    rawlsDAO.getWorkspace(workspaceNamespace, workspaceName)(userInfo).map(_.workspace.workspaceId)
+  }
+
+  private def importToCWDS(workspaceNamespace: String, workspaceName: String, workspaceId: String, userInfo: UserInfo, importRequest: AsyncImportRequest
                           ): Future[PerRequestMessage] = {
-      // translate the workspace namespace/name into an id
-      rawlsDAO.getWorkspace(workspaceNamespace, workspaceName)(userInfo) map { workspace =>
-      // create the job in cWDS
-      val cwdsJob = cwdsDAO.importV1(workspace.workspace.workspaceId, importRequest)(userInfo)
-      // massage the cWDS job into the response format Orch requires
-      val asyncImportResponse = AsyncImportResponse(url = importRequest.url,
+    // create the job in cWDS
+    val cwdsJob = cwdsDAO.importV1(workspaceId, importRequest)(userInfo)
+    // massage the cWDS job into the response format Orch requires
+    Future.successful(
+      RequestComplete(Accepted, AsyncImportResponse(url = importRequest.url,
         jobId = cwdsJob.getJobId.toString,
-        workspace = WorkspaceName(workspaceNamespace, workspaceName))
-      RequestComplete(Accepted, asyncImportResponse)
-    }
+        workspace = WorkspaceName(workspaceNamespace, workspaceName)))
+    )
   }
```

Comment:
now that `importToCWDS` no longer includes `getWorkspace`, there's no need for it to return a `Future`. It can be simplified by removing `Future[...]` from its return type and by removing `Future.successful(…)` from its implementation:
```scala
  private def importToCWDS(workspaceNamespace: String, workspaceName: String, workspaceId: String, userInfo: UserInfo, importRequest: AsyncImportRequest
                          ): PerRequestMessage = {
    // create the job in cWDS
    val cwdsJob = cwdsDAO.importV1(workspaceId, importRequest)(userInfo)
    // massage the cWDS job into the response format Orch requires
    RequestComplete(Accepted, AsyncImportResponse(url = importRequest.url,
      jobId = cwdsJob.getJobId.toString,
      workspace = WorkspaceName(workspaceNamespace, workspaceName)))
  }
```

then, in the two places it is called, you can change `getWorkspaceId … flatMap … importToCWDS` to `getWorkspaceId … map … importToCWDS`

Reply:
Nice!!!